### PR TITLE
Phoenix third module

### PR DESCRIPTION
**Pull request by samu975**

### CODE REVIEW

**Comment by juanhurtado10** [comment]
@samu975 Could you please merge #10 and then rebase this branch? Thanks!

File: Phoenix/hello/lib/hello/catalog.ex [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+      iex> get_product!(123)
+      %Product{}
+
+      iex> get_product!(456)
+      ** (Ecto.NoResultsError)
+
+  """
+  def get_product!(id) do
+    Product |> Repo.get(id) |> Repo.preload(:categories)
```
Comment (juanhurtado10) [inline Phoenix/hello/lib/hello/catalog.ex:39]: @samu975 what does a preload do?

File: Phoenix/hello/lib/hello/catalog.ex [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+  @doc """
+  Creates a product.
+
+  ## Examples
+
+      iex> create_product(%{field: value})
+      {:ok, %Product{}}
+
+      iex> create_product(%{field: bad_value})
+      {:error, %Ecto.Changeset{}}
```
Comment (juanhurtado10) [inline Phoenix/hello/lib/hello/catalog.ex:50]: @samu975 could you explain how a changeset works?

File: Phoenix/hello/lib/hello/catalog/category.ex [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  schema "categories" do
+    field :title, :string
+
+    timestamps()
+  end
+
+  @doc false
+  def changeset(category, attrs) do
+    category
+    |> cast(attrs, [:title])
```
Comment (juanhurtado10) [inline Phoenix/hello/lib/hello/catalog/category.ex:14]: @samu975 could you explain what this line does?

File: Phoenix/hello/lib/hello/shopping_cart.ex [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+      {:error, changeset} -> {:error, changeset}
+    end
+  end
+
+  defp reload_cart(%Cart{} = cart), do: get_cart_by_user_uuid(cart.user_uuid)
+
+  def add_item_to_cart(%Cart{} = cart, %Catalog.Product{} = product) do
+    %CartItem{quantity: 1, price_when_carted: product.price}
+    |> CartItem.changeset(%{})
+    |> Ecto.Changeset.put_assoc(:cart, cart)
```
Comment (juanhurtado10) [inline Phoenix/hello/lib/hello/shopping_cart.ex:96]: @samu975 what is an association?

File: Phoenix/hello/lib/hello_web/templates/hello/show.html.heex [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<section class="phx-hero">
+    <h2>Hello World, from <%= @messenger %>!</h2>
+</section>
```
Comment (juanhurtado10) [inline Phoenix/hello/lib/hello_web/templates/hello/show.html.heex:3]: @samu975 Again, please configure your text editor to automatically add a blank line at the end of the files

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: Phoenix/hello/lib/hello/catalog/category.ex (hunk 3 above)
Comment (samu975) [reply Phoenix/hello/lib/hello/catalog/category.ex:14]: Cast takes three parameters. First a struct, in this case I think is "category". Second some attributes, in this case, attrs. And the last one is a list of attributes that are allowed to be altered, in this case is the title. These lines prevent some fields from being updated.

File: Phoenix/hello/lib/hello/catalog.ex (hunk 2 above)
Comment (samu975) [reply Phoenix/hello/lib/hello/catalog.ex:50]: changeset is used for type-casting, user input validation, and filtering parameters. Is a struct that stores a set of changes. We can use cast, validate_required, validate_format, validate_inclusion, etc.

File: Phoenix/hello/lib/hello/catalog.ex (hunk 1 above)
Comment (samu975) [reply Phoenix/hello/lib/hello/catalog.ex:39]: Preload, as its name says serves to preload, in this case, the categories. It allows us to reference them in templates, controllers, and views.

File: Phoenix/hello/lib/hello_web/templates/hello/show.html.heex (hunk 5 above)
Comment (samu975) [reply Phoenix/hello/lib/hello_web/templates/hello/show.html.heex:3]: I already did, maybe these files was created before of config the vs code. Sorry about that

File: Phoenix/hello/lib/hello/shopping_cart.ex (hunk 4 above)
Comment (samu975) [reply Phoenix/hello/lib/hello/shopping_cart.ex:96]: An association is used to associate data with other data. There are three types of associations. has_many, to say that data belongs to other data or has several of some data. belong_to, to associate one data is related with another. has_one, to say that x data type has a data type.

File: Phoenix/hello/lib/hello/catalog.ex (hunk 1 above)
Comment (juanhurtado10) [reply Phoenix/hello/lib/hello/catalog.ex:39]: 🤔 seems it is not clear, please revisit the guides.

**Review by juanhurtado10 — APPROVED** [review]

**Pull request merged**